### PR TITLE
Channel System is Expanded!

### PR DESCRIPTION
**Pull request by cydh**
* **Addressed Issue(s)**: 
  * #1886 
  * #1342

* **Server Mode**: Both

* **Description of Pull Request**: 
  * Base: https://github.com/cydh/rathena/pull/2
  * Many changes on conf/channel.conf!
  * Many new channel script commands!
  * Grammar correction by @aleos89 as always!
  * Forum: https://rathena.org/board/topic/107823-channel-system-expanded/

### CODE REVIEW

**Comment by RadianFord** [comment]
```
channel.c: In function ‘channel_pcleave’:
channel.c:718:8: warning: declaration of ‘output’ shadows a previous local [-Wshadow]
   char output[CHAT_SIZE_MAX+1];
        ^
channel.c:694:7: note: shadowed declaration is here
  char output[128];
       ^
channel.c: In function ‘channel_pckick’:
channel.c:1060:8: warning: declaration of ‘output’ shadows a previous local [-Wshadow]
   char output[CHAT_SIZE_MAX+1];
        ^
channel.c:1017:7: note: shadowed declaration is here
  char output[CHAT_SIZE_MAX];
       ^
channel.c: In function ‘channel_pcisjoined_sub’:
channel.c:1253:25: warning: cast from pointer to integer of different size [-Wpointer-to-int-cast]
  unsigned int char_id = (unsigned int)db_data2ptr(data);
                         ^
channel.c: In function ‘channel_read_sub’:
channel.c:1301:50: warning: passing argument 3 of ‘config_setting_lookup_string’ from incompatible pointer type [-Wincompatible-pointer-types]
  if (!config_setting_lookup_string(chan, "name", &name)) {
                                                  ^
In file included from ../common/conf.h:8:0,
                 from channel.c:6:
../common/../../3rdparty/libconfig/libconfig.h:160:26: note: expected ‘const char **’ but argument is of type ‘char **’
 extern LIBCONFIG_API int config_setting_lookup_string(
                          ^
channel.c:1305:49: warning: passing argument 3 of ‘config_setting_lookup_string’ from incompatible pointer type [-Wincompatible-pointer-types]
  if (config_setting_lookup_string(chan, "type", &type_str) && !script_get_const
                                                 ^
In file included from ../common/conf.h:8:0,
                 from channel.c:6:
../common/../../3rdparty/libconfig/libconfig.h:160:26: note: expected ‘const char **’ but argument is of type ‘char **’
 extern LIBCONFIG_API int config_setting_lookup_string(
                          ^
channel.c:1309:49: warning: passing argument 3 of ‘config_setting_lookup_string’ from incompatible pointer type [-Wincompatible-pointer-types]
  config_setting_lookup_string(chan, "password", &password);
                                                 ^
In file included from ../common/conf.h:8:0,
                 from channel.c:6:
../common/../../3rdparty/libconfig/libconfig.h:160:26: note: expected ‘const char **’ but argument is of type ‘char **’
 extern LIBCONFIG_API int config_setting_lookup_string(
                          ^
channel.c:1310:46: warning: passing argument 3 of ‘config_setting_lookup_string’ from incompatible pointer type [-Wincompatible-pointer-types]
  config_setting_lookup_string(chan, "alias", &alias);
                                              ^
In file included from ../common/conf.h:8:0,
                 from channel.c:6:
../common/../../3rdparty/libconfig/libconfig.h:160:26: note: expected ‘const char **’ but argument is of type ‘char **’
 extern LIBCONFIG_API int config_setting_lookup_string(
                          ^
channel.c:1311:46: warning: passing argument 3 of ‘config_setting_lookup_string’ from incompatible pointer type [-Wincompatible-pointer-types]
  config_setting_lookup_string(chan, "color", &color_str);
                                              ^
In file included from ../common/conf.h:8:0,
                 from channel.c:6:
../common/../../3rdparty/libconfig/libconfig.h:160:26: note: expected ‘const char **’ but argument is of type ‘char **’
 extern LIBCONFIG_API int config_setting_lookup_string(
                          ^
channel.c:1291:6: warning: unused variable ‘id’ [-Wunused-variable]
  int id = 0;
      ^
channel.c: In function ‘channel_read_config’:
channel.c:1363:37: error: ‘channel_conf’ undeclared (first use in this function)
  if (conf_read_file(&channels_conf, channel_conf)) {
                                     ^
channel.c:1363:37: note: each undeclared identifier is reported only once for each function it appears in
channel.c:1370:21: warning: declaration of ‘colors’ shadows a previous local [-Wshadow]
   config_setting_t *colors, *private_channel = NULL, *channels = NULL;
                     ^
channel.c:1360:20: note: shadowed declaration is here
  config_setting_t *colors = NULL;
                    ^
channel.c:1399:59: warning: passing argument 3 of ‘config_setting_lookup_string’ from incompatible pointer type [-Wincompatible-pointer-types]
    config_setting_lookup_string(private_channel, "color", &color_str);
                                                           ^
In file included from ../common/conf.h:8:0,
                 from channel.c:6:
../common/../../3rdparty/libconfig/libconfig.h:160:26: note: expected ‘const char **’ but argument is of type ‘char **’
 extern LIBCONFIG_API int config_setting_lookup_string(
                          ^
channel.c:1360:20: warning: unused variable ‘colors’ [-Wunused-variable]
  config_setting_t *colors = NULL;
                    ^
channel.c:1359:20: warning: unused variable ‘chsys’ [-Wunused-variable]
  config_setting_t *chsys = NULL;
```
I add the changes manually.. and warnings errors show up

File: src/map/clif.c [hunk 1]
```diff
@@ -6190,31 +6190,35 @@ void clif_broadcast2(struct block_list* bl, const char* mes, int len, unsigned l
...
- * Display *msg from *sd to all *users in channel
+ * Display *msg to all *users in channel
  */
-void clif_channel_msg(struct Channel *channel, struct map_session_data *sd, char *msg, short color) {
+void clif_channel_msg(struct Channel *channel, const char *msg, unsigned long color) {
 	DBIterator *iter;
 	struct map_session_data *user;
-	unsigned short msg_len = (unsigned short)(strlen(msg) + 1);
+	unsigned short msg_len = 0, len = 0;
+	unsigned char buf[12+CHAT_SIZE_MAX+1];
```
Comment (anacondaqq) [inline src/map/clif.c]: is not this provoke overflow? CHAT_SIZE_MAX = 255 + 1

**Comment by Vexillio** [comment]
** Im having erros when compiling @.@
`Error	1	error C2065: 'buyingstores_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\buyingstore.c	231	1	map-server
Error	2	error C2065: 'buyingstore_items_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\buyingstore.c	236	1	map-server
Error	3	error C2065: 'buyingstore_items_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\buyingstore.c	262	1	map-server
Error	4	error C2065: 'buyingstores_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\buyingstore.c	263	1	map-server
Error	5	error C2065: 'buyingstore_items_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\buyingstore.c	455	1	map-server
Error	6	error C2065: 'buyingstore_items_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\buyingstore.c	459	1	map-server
Error	7	error C2065: 'buyingstores_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\buyingstore.c	491	1	map-server
Error	8	error C2065: 'buyingstores_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\buyingstore.c	659	1	map-server
Error	9	error C2065: 'buyingstore_items_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\buyingstore.c	659	1	map-server
Error	10	error C2065: 'buyingstore_items_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\buyingstore.c	715	1	map-server
Error	11	error C2065: 'buyingstores_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\buyingstore.c	750	1	map-server
Error	12	error C2065: 'buyingstore_items_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\buyingstore.c	751	1	map-server
Error	13	error C2065: 'roulette_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\itemdb.c	1072	1	map-server
Error	14	error C2065: 'roulette_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\itemdb.c	1152	1	map-server
Error	15	error C2065: 'item_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\itemdb.c	1493	1	map-server
Warning	16	warning C4047: 'initializing' : 'const char *' differs in levels of indirection from 'int'	c:\users\vexillio\desktop\rathena-master\src\map\itemdb.c	1493	1	map-server
Error	17	error C2065: 'item2_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\itemdb.c	1495	1	map-server
Warning	18	warning C4047: 'initializing' : 'const char *' differs in levels of indirection from 'int'	c:\users\vexillio\desktop\rathena-master\src\map\itemdb.c	1495	1	map-server
Error	19	error C2065: 'mob_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\mob.c	4175	1	map-server
Warning	20	warning C4047: 'initializing' : 'const char *' differs in levels of indirection from 'int'	c:\users\vexillio\desktop\rathena-master\src\map\mob.c	4175	1	map-server
Error	21	error C2065: 'mob2_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\mob.c	4176	1	map-server
Warning	22	warning C4047: 'initializing' : 'const char *' differs in levels of indirection from 'int'	c:\users\vexillio\desktop\rathena-master\src\map\mob.c	4176	1	map-server
Error	23	error C2065: 'mob_skill_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\mob.c	4635	1	map-server
Warning	24	warning C4047: 'initializing' : 'const char *' differs in levels of indirection from 'int'	c:\users\vexillio\desktop\rathena-master\src\map\mob.c	4635	1	map-server
Error	25	error C2065: 'mob_skill2_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\mob.c	4636	1	map-server
Warning	26	warning C4047: 'initializing' : 'const char *' differs in levels of indirection from 'int'	c:\users\vexillio\desktop\rathena-master\src\map\mob.c	4636	1	map-server
Error	27	error C2065: 'OPT1_CRYSTALIZE' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\pc.c	5228	1	map-server
Error	28	error C2065: 'SC_ARMOR_ELEMENT' : undeclared identifier (..\src\map\script.c)	c:\users\vexillio\desktop\rathena-master\src\map\script_constants.h	879	1	map-server
Error	29	error C2065: 'vending_items_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\vending.c	54	1	map-server
Error	30	error C2065: 'vendings_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\vending.c	55	1	map-server
Error	31	error C2065: 'vending_items_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\vending.c	210	1	map-server
Error	32	error C2065: 'vending_items_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\vending.c	214	1	map-server
Error	33	error C2065: 'vendings_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\vending.c	350	1	map-server
Error	34	error C2065: 'vending_items_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\vending.c	355	1	map-server
Error	35	error C2065: 'vendings_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\vending.c	541	1	map-server
Error	36	error C2065: 'vending_items_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\vending.c	541	1	map-server
Error	37	error C2065: 'vending_items_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\vending.c	596	1	map-server
Error	38	error C2065: 'vendings_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\vending.c	631	1	map-server
Error	39	error C2065: 'vending_items_table' : undeclared identifier	c:\users\vexillio\desktop\rathena-master\src\map\vending.c	632	1	map-server
`

**Comment by aleos89** [comment]
@Vexillio: That's a conflict with your repo. The tables were changed in 5d22e1c. It's not related to this PR.

**Comment by RadianFord** [comment]
I got errors while testing the `*channel_chat "<chname>","<message>"{,<color>};`
```
[Warning]: Unexpected type for argument 1. Expected string.
[Debug]: Data: variable name='#global' index=0
[Debug]: Function: channel_chat
[Debug]: Source (NPC): Channel_Test (invisible/not on a map)
[Error]: script_rid2sd: fatal error ! player not attached!
[Debug]: Function: channel_chat (2 parameters):
[Debug]: Data: variable name='#global' index=0
[Debug]: Data: string value="Test Channel for Global!"
[Debug]: Source (NPC): Channel_Test (invisible/not on a map)
[Warning]: script:get_val: cannot access player variable '#global', defaulting to 0
[Debug]: (buildin_channel_chat) mapindex_name2id: Map "0" not found in index list!
[Error]: buildin_channel_chat: Invalid map '0'.
[Debug]: Source (NPC): Channel_Test (invisible/not on a map)
```

My script
```
-	script	Channel_Test	FAKE_NPC,{
	end;
	OnMinute30:
	switch(rand(1,2)) {
			case 1: channel_chat(#event,"Test Channel for Event!"); end;
			case 2: channel_chat(#global,"Test Channel for Global!"); end;
		}
	end;
}
``

**Comment by Akkarinage** [comment]
Command description: channel_chat **"**<chname>**"**,"<message>"{,<color>};
You're using: channel_chat(#event,"Test Channel for Event!");
You should be using: channel_chat("#event","Test Channel for Event!");

**Comment by RadianFord** [comment]
@Akkarinage oh it worked thanks!

**Comment by Vexillio** [comment]
It seems im having an map crash whenever i do @autotrade in my server with a fresh rA file and only have a channel system modifications. can someone confirm this? so i can know if im the only one having this error.

In the hosted server im having this error.
[Error]: Server received crash signal! attempting to save all online characters!
[Error]: pc_eventtimer: no such event timer

On a local host im having a map crash.

**Comment by anacondaqq** [comment]
@Vexillio hi, will be super nice if you provide complete crash report. If are you using linux -> try to complile server with gdb support. If you are using Visual Studio 2010+ try to complile server in debug mode, and when it will crash -> open crash in VS and show lines which provoke crash.

**Comment by Vexillio** [comment]
@anacondaqq Im using 2012 VS and it seems that whenever i did debug-start debug only the char server is running and btw heres the screenshot of the error on the host i hope this helps.
http://imgur.com/a/ptRmP
The saddest part the server seems not saving data's example a player got apple after we restart the server or the player character select the apple is missing.

**Comment by Vexillio** [comment]
@anacondaqq I did what you told me just now, and i got no errors when i do normal recompile on windows clean>build solution no errors but after that i did clean>code analysis i got 250 warnings char-server(81 warning), login-server (41 warnings), mapcache (11 warnings), map-server (117 warnings)

Here's the screenshot when i did normal compilation on windows VS 2012:
http://i.imgur.com/W5qPewt.png

Here's the screenshot when i run code analysis on windows VS 2012:
http://i.imgur.com/uSmE1fI.png

**Comment by Vexillio** [comment]
@anacondaqq latest update i already managed to fix the errors thanks to @RadianFord the only thing now is i can compile and run the server on my localhost im using windows but on my host i cant compile here's the error log.

`make[1]: Entering directory `/home/ros/ro/3rdparty/mt19937ar'
make[1]: Nothing to be done for `all'.
make[1]: Leaving directory `/home/ros/ro/3rdparty/mt19937ar'
make[1]: Entering directory `/home/ros/ro/3rdparty/libconfig'
make[1]: Nothing to be done for `all'.
make[1]: Leaving directory `/home/ros/ro/3rdparty/libconfig'
make[1]: Entering directory `/home/ros/ro/src/common'
make[1]: Warning: File `core.c' has modification time 6.8e+04 s in the future
        CC      core.c
        CC      socket.c
        CC      timer.c
        CC      db.c
        CC      nullpo.c
        CC      malloc.c
        CC      showmsg.c
        CC      strlib.c
        CC      utils.c
        CC      grfio.c
        CC      mapindex.c
        CC      ers.c
        CC      md5calc.c
        CC      core.c
        CC      socket.c
        CC      malloc.c
        CC      random.c
        CC      des.c
        CC      conf.c
        CC      thread.c
        CC      mutex.c
        CC      raconf.c
        CC      mempool.c
        CC      msg_conf.c
        CC      cli.c
        CC      sql.c
        AR      obj/common.a
make[1]: warning:  Clock skew detected.  Your build may be incomplete.
make[1]: Leaving directory `/home/ros/ro/src/common'
make[1]: Entering directory `/home/ros/ro/src/login'
make[1]: Warning: File `account.c' has modification time 6.8e+04 s in the future
        CC      account.c
        CC      ipban.c
        CC      login.c
        CC      loginchrif.c
        CC      loginclif.c
        CC      logincnslif.c
        CC      loginlog.c
        LD      login-server
make[1]: warning:  Clock skew detected.  Your build may be incomplete.
make[1]: Leaving directory `/home/ros/ro/src/login'
make[1]: Entering directory `/home/ros/ro/src/char'
make[1]: Warning: File `char.c' has modification time 6.8e+04 s in the future
        CC      char.c
        CC      char_clif.c
        CC      char_cnslif.c
        CC      char_logif.c
        CC      char_mapif.c
        CC      int_auction.c
        CC      int_clan.c
        CC      int_elemental.c
        CC      inter.c
        CC      int_guild.c
        CC      int_homun.c
        CC      int_mail.c
        CC      int_mercenary.c
        CC      int_party.c
        CC      int_pet.c
        CC      int_quest.c
        CC      int_storage.c
        LD      char-server
make[1]: warning:  Clock skew detected.  Your build may be incomplete.
make[1]: Leaving directory `/home/ros/ro/src/char'
make[1]: Entering directory `/home/ros/ro/src/map'
make[1]: Warning: File `../map/atcommand.h' has modification time 6.8e+04 s in the future
        CC      atcommand.c
        CC      battle.c
        CC      battleground.c
        CC      buyingstore.c
        CC      cashshop.c
        CC      channel.c
channel.c: In function ‘channel_pcisjoined_sub’:
channel.c:1252: warning: cast from pointer to integer of different size
channel.c: In function ‘channel_read_sub’:
channel.c:1300: warning: passing argument 3 of ‘config_setting_lookup_string’ from incompatible pointer type
../common/../../3rdparty/libconfig/libconfig.h:160: note: expected ‘const char ’ but argument is of type ‘char ’
channel.c:1304: warning: passing argument 3 of ‘config_setting_lookup_string’ from incompatible pointer type
../common/../../3rdparty/libconfig/libconfig.h:160: note: expected ‘const char ’ but argument is of type ‘char ’
channel.c:1308: warning: passing argument 3 of ‘config_setting_lookup_string’ from incompatible pointer type
../common/../../3rdparty/libconfig/libconfig.h:160: note: expected ‘const char ’ but argument is of type ‘char ’
channel.c:1309: warning: passing argument 3 of ‘config_setting_lookup_string’ from incompatible pointer type
../common/../../3rdparty/libconfig/libconfig.h:160: note: expected ‘const char ’ but argument is of type ‘char ’
channel.c:1310: warning: passing argument 3 of ‘config_setting_lookup_string’ from incompatible pointer type
../common/../../3rdparty/libconfig/libconfig.h:160: note: expected ‘const char ’ but argument is of type ‘char ’
channel.c:1290: warning: unused variable ‘id’
channel.c: In function ‘channel_read_config’:
channel.c:1398: warning: passing argument 3 of ‘config_setting_lookup_string’ from incompatible pointer type
../common/../../3rdparty/libconfig/libconfig.h:160: note: expected ‘const char ’ but argument is of type ‘char ’
channel.c:1359: warning: unused variable ‘colors’
channel.c:1358: warning: unused variable ‘chsys’
        CC      chat.c
        CC      chrif.c
        CC      clan.c
        CC      clif.c
        CC      date.c
        CC      duel.c
        CC      elemental.c
        CC      guild.c
        CC      homunculus.c
        CC      instance.c
        CC      intif.c
        CC      itemdb.c
        CC      log.c
        CC      mail.c
        CC      map.c
        CC      mapreg.c
        CC      mercenary.c
        CC      mob.c
        CC      npc.c
        CC      npc_chat.c
        CC      party.c
        CC      path.c
        CC      pc.c
        CC      pc_groups.c
        CC      pet.c
        CC      quest.c
        CC      script.c
        CC      searchstore.c
        CC      skill.c
        CC      status.c
        CC      storage.c
        CC      trade.c
        CC      unit.c
        CC      vending.c
        LD      map-server
make[1]: warning:  Clock skew detected.  Your build may be incomplete.
make[1]: Leaving directory `/home/ros/ro/src/map'
building conf/import, conf/msg_conf/import and db/import folder...
[666ros@njded242 ro`

**Comment by anacondaqq** [comment]
@Vexillio 
> [Error]: pc_eventtimer: no such event timer

I got the same issue earlier some time ago on the very customized emulator, not clean rAthena. The issue were related to d99c91418ac671ca8a46d53d226f40164f7139d7 and my custom command (detach) which leave player online like in autotrade mode while player in the game ([modified example](https://rathena.org/board/topic/71389-afk-with-afk-hat/)) and problem were with `chrif_save(sd, SOME NUMBER HERE);` and after a little bit analyzing and finding how to use new chrif_save, I changed it to proper values (`chrif_save(sd, CSAVE_AUTOTRADE);`) and my problem has gone.

Try to search in all files for chrif_save(sd, NUMBER VALUE) in your sources, and replace with constants) like in commit above.

### So, in your case what you need to do, it's 

1. Compare autotrade function inside atcommand.c in your emulator with autotradr function in atcommand.c in rAthena, if it different -> makes them the same.

2. If it doesn't help, then you need realise what a reason of your map-server crash. For that, you should use debugging and when you got a crash -> need to report about a full trace of the problem.

### So what you need to do now:

As I understand on the screen above you are using Putty + some linux distro where is run server.
If the server still crashing time to time, it's time to compile and run the server in debug mode.
Before switching to debug mode, few theoretical details about what you should get in result:

1. you install special debugging software for debugger
2. then you compile the server in debug mode
3. then you need to run server in debugger in separate window
4. when your server will crash -> it will freeze and you just open window with debugger on your linux, and copy-paste full log of crash here
5. such info will be helpful to understand is it you have a problem, or it's problem with the code.

## So here we go:

1. install gdb first (depends on your distro), I prefer ubuntu / Debian like distros, so my command will be next:

> apt-get install gdb screen

Where gdb = debugger, screen = tool for running each process in separate terminal (for easy switching between login / char / map-server)

2. Then, let's clean old code and compilation garbage and try to compile our server with gdb support (gdb = debugger) for that we need to do next:

**a)** reboot your linux server
**b)** go to your RO folder with `cd /home/ros/ro/`
**c)** then make permissions for your configure file

> chmod +x configure

**d)** then clean your project via

> make clean

**e)** then try to compile it again with gdb support, then wait when compilation will be finished

> ./configure --enable-debug=gdb

**f)** then, we need run debugger (gdb) and inside this debugger we will attach our RO server (map-server) which as I understood crashing, but for that we need to use screen (for running each server in separate terminal) (you should be inside your linux server folder for running next commands)

**I.**

> screen -amdS loginserver ./login-server

**II.**

> screen -amdS charserver ./char-server

The first two servers right now running in separate screen terminals, you can always see what is going on with login or char server via  `screen -ls` and attach terminal via `screen -X loginserver/charserver` and detach from terminal via `CTRL+A, then press D`

**III.**

Right now we will open separate terminal in screen and run gdb than in gdb we will restart our app

#### So:

a.) create separate screen terminal via screen -S "mapserv"
b.) go inside this terminal via screen -X mapserv
c.) do commands below:

> gdb map-sersver

it will attach to GDB map-server and you will see gdb window
inside this window just enter "r" and press enter.

It will restart attached to gdb map-server and you will see that server in running.
Detach from screen terminal via `CTRL + A, then press D`

So, you finished, try to check the connection to your server. And try to reproduce your bug.
If map-server will crash, you will see "freeze" in the game, when nothing will move / or react.
That means -> map-server maybe crashed and you need to check it, and if-if crashed -> post full details (trace) about the crash here with lines in your code which provoke a crash.

### So, your actions are next:

1. Your server freeze
2. You are open your ssh terminal
3. You attaching to your previously created mapserv screen terminal via `screen -X mapserv`
4. You see gdb output that your server crashed.
5. You copy-paste each of #0 up to #10 or something like that number output and paste it here or to pastebin.


## Important notes


1. I suggest you do hard way, because not always gdb can provide output because of different settings in your system with secure.limits. For sure you can try to read and try to use other method (but for the same result) at rAthena boards. Just use next link: [How to use GDB](http://bfy.tw/AHJK)

2. I do not talk anything about Visual Studio 2012 and Windows because it will be a little bit harder to find out where is a problem, and to be clear i' m very lazy for explaining official documentation from visual studio which you can find on the internet.

**Comment by RadianFord** [comment]
Another Issue in game

I did set delay for the channel main now if i use a level 99 account of course it can bypass the settings but if a normal player use/type in the channel main the message You're talking to fast! is triggered even tho I just got login in game.

**Comment by admkakaroto** [comment]
@RadianFord  I'm also having the same problem.

**Comment by RadianFord** [comment]
@admkakaroto No.. still having issue with it I hope they'll update it soon

**Comment by admkakaroto** [comment]
Yeah ... it's complicated.

**Comment by aleos89** [comment]
@RadianFord @admkakaroto I'm unable to reproduce the issue with channel delays. All my channels are defaulted to the 1000ms delay and I can chat normally in every channel. If I start to spam I get the "You're talking too fast!" message and continue chatting normally after.
Here's a private channel test:
image
Here's a public channel test:
image

**Comment by admkakaroto** [comment]
@aleos89 If you send the same message you will be able to reproduce the problem, if sending a different message works normally.

**Comment by admkakaroto** [comment]
@cydh If you send the same message you will be able to reproduce the problem, if sending a different message works normally.

**Comment by aleos89** [comment]
Just sent "test" 7 times in a row.
image

**Comment by admkakaroto** [comment]
If delay: 1000 ?

**Comment by aleos89** [comment]
Yes, I left the channels.conf with all default values.

**Comment by admkakaroto** [comment]
So I'm going to apply again to the clean rathena to test ...

**Comment by admkakaroto** [comment]
@cydh http://i.imgsafe.org/1282b91d41.png

**Comment by admkakaroto** [comment]
I can send 2 times in a row and it totally ignores the delay, then displays the message blocking message but not correctly, and also does not release to send another message, unless I send the different message, but if it sends the same message It will not allow you to send another message.

**Comment by aleos89** [comment]
That message is the generic message sent by the client to stop spam. That has nothing to do with the channel system.

**Comment by admkakaroto** [comment]
Yes I know, it happens that it does not release the player to send another message to the channel system again, so he has to send a different message, otherwise the delay does not work.

**Comment by admkakaroto** [comment]
I'm going to do a little test to see if it works out.

**Comment by admkakaroto** [comment]
I applied 'Chat Flood Remove Limit' with Nemo Patcher, and the channel system is not adding any delay messages.

http://i.imgsafe.org/12b6a5e8e1.png

**Comment by Lemongrass3110** [comment]
This is because it is a clientside protection message that has nothing to do with this feature at all...

**Comment by admkakaroto** [comment]
Yes I know this, see in the image, the channel delay is not being applied to the player, I am with a default configuration of channel.conf and it does not work delay, see:

http://i.imgsafe.org/12b6a5e8e1.png

**Comment by Lemongrass3110** [comment]
Are you testing on rAthena or rAmod again?

**Comment by admkakaroto** [comment]
@Lemongrass3110 Rathena, I do not use rAmod anymore, just rAthena, totally clean, downloaded less than 2 hours from github.

**Comment by RadianFord** [comment]
@aleos89 Is that a normal account not a GM?

@admkakaroto what client date are you using? its not RE client its ragexe right? Im using this client 2015-05-13aRagexe

**Comment by aleos89** [comment]
@RadianFord: Yes, I made sure to login on a non-GM account.

**Comment by RadianFord** [comment]
@aleos89 What client date are you using? and is the packet obfuscation active? is it renewal mode?

**Comment by aleos89** [comment]
It shouldn't matter upon the client or server mode. I'm using 2015-10-29 with packet obfuscation and renewal mode.

**Comment by admkakaroto** [comment]
I use 2015-11-04 without obsfucation, renewal mode.

**Comment by RadianFord** [comment]
@aleos89 oh.. then i dont have anything to say what went wrong with my files.

**Pull request merged**